httpbakery/agent/agent.go

```diff
@@ -0,0 +1,154 @@
+/*
+Package agent implements the agent login mechanism. The agent login
+mechanism provides a form of interaction that can be used in
+situations where interaction with the user is not possible.
+
+An agent login works as follows:
+
+	    Agent                            Login Service
+	      |                                    |
+	      | GET visitURL with agent cookie     |
+	      |----------------------------------->|
+	      |                                    |
+	      |    Macaroon with local third-party | 
+	      |                 discharge required |
+	      |<-----------------------------------|
+	      |                                    |
+	      | GET visitURL with agent cookie &   |
+	      | discharged macaroon                |
+	      |----------------------------------->|
+	      |                                    |
+	      |               Agent login response |
+	      |<-----------------------------------|
+	      |                                    |
+
+The agent cookie is a cookie named "agent-login" its value is a base64
+encoded JSON object of the form:
+
+	{
+		"username": <username>,
+		"public_key": <public key>
+	}
+
+The httpbakery.Client handles the local third-party discharge.
+
+The agent login response will be a JSON object of the same form as the agent cookie.
+*/
+package agent
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"net/url"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+
+	"gopkg.in/macaroon-bakery.v1/bakery"
+	"gopkg.in/macaroon-bakery.v1/httpbakery"
+)
+
+var logger = loggo.GetLogger("httpbakery.agent")
+
+// agentLogin contains the structure of an agent login cooke. It is also
+// returned in a successful agent login attempt to help indicate that an
+// agent login has occurred.
+type agentLogin struct {
+	Username  string            `json:"username"`
+	PublicKey *bakery.PublicKey `json:"public_key"`
+}
+
+// SetCookie creates a cookie in c which is suitable for performing agent
+// logins to u. The username will be combined with the public key taken
+// from c.Key.Public, c.Key must have been set before calling SetCookie.
+// If c is used with VisitWebPage interactions with u will be performed as
+// an agent.
+func SetCookie(c *httpbakery.Client, u *url.URL, username string) {
+	al := agentLogin{
+		Username:  username,
+		PublicKey: &c.Key.Public,
+	}
+	b, err := json.Marshal(al)
+	if err != nil {
+		// This shouldn't happen as the agentLogin type has to be marshalable.
+		panic(errgo.Notef(err, "cannot marshal cookie"))
```

Comment:
Better to do this than to force each caller to check for an error that
can actually never happen, I think.